packages/react-dom/src/fire/ReactFireComponentProperties.js

```diff
@@ -0,0 +1,905 @@
+      const lastChildren = lastRawProps.children;
+      if (
+        typeof lastChildren !== 'string' ||
+        domNode.firstChild == null ||
+        propValue === ''
+      ) {
+        domNode.textContent = propValue;
+      } else {
+        domNode.firstChild.nodeValue = propValue;
+      }
+    } else if (isPropAnEvent(propName) && !isCustomComponentTag) {
+      setEventProp(propName, propValue, domNode);
+    } else {
+      setValueForProperty(domNode, propName, propValue, isCustomComponentTag);
+    }
+  }
+}
+
+export function diffDOMElementProperties(
+  domNode: Element,
+  type: string,
+  updatePayload: null | Array<any>,
+  lastProps: Object,
+  nextProps: Object,
+  rootContainerElement: Element | Document,
+): null | Array<any> {
+  assertValidProps(type, nextProps);
+
+  let propName;
+  let styleName;
+  let styleUpdates = null;
+  for (propName in lastProps) {
+    if (
+      nextProps.hasOwnProperty(propName) ||
+      !lastProps.hasOwnProperty(propName) ||
+      lastProps[propName] == null
+    ) {
+      continue;
+    }
+    if (propName === STYLE) {
+      const lastStyle = lastProps[propName];
+      for (styleName in lastStyle) {
+        if (lastStyle.hasOwnProperty(styleName)) {
+          if (!styleUpdates) {
+            styleUpdates = {};
+          }
+          styleUpdates[styleName] = '';
+        }
+      }
+    } else if (
+      propName === DANGEROUSLY_SET_INNER_HTML ||
+      propName === CHILDREN
+    ) {
+      // Noop. This is handled by the clear text mechanism.
+    } else if (
+      propName === SUPPRESS_CONTENT_EDITABLE_WARNING ||
+      propName === SUPPRESS_HYDRATION_WARNING
+    ) {
+      // Noop
+    } else if (propName === AUTOFOCUS) {
+      // Noop. It doesn't work on updates anyway.
+    } else {
+      // For all other deleted properties we add it to the queue. We use
+      // the whitelist in the commit phase instead.
+      (updatePayload = updatePayload || []).push(propName, null);
+    }
+  }
+  for (propName in nextProps) {
+    const nextProp = nextProps[propName];
+    const lastProp = lastProps != null ? lastProps[propName] : undefined;
+    if (
+      !nextProps.hasOwnProperty(propName) ||
+      nextProp === lastProp ||
+      (nextProp == null && lastProp == null)
+    ) {
+      continue;
+    }
+    if (propName === STYLE) {
+      if (__DEV__) {
+        if (nextProp) {
+          // Freeze the next style object so that we can assume it won't be
+          // mutated. We have already warned for this in the past.
+          Object.freeze(nextProp);
+        }
+      }
+      if (lastProp) {
+        // Unset styles on `lastProp` but not on `nextProp`.
+        for (styleName in lastProp) {
+          if (
+            lastProp.hasOwnProperty(styleName) &&
+            (!nextProp || !nextProp.hasOwnProperty(styleName))
+          ) {
+            if (!styleUpdates) {
+              styleUpdates = {};
+            }
+            styleUpdates[styleName] = '';
```

Comment:
it looks like repeating code from line 306. Is it possible to simplify logic here, maybe to extract method?